### ISSUE TITLE
does the simam attention needs `detach`?

### ISSUE THREAD

**Issue by guozhiyao** [ZjjConan/SimAM#3]
Does the attention value `self.activaton(y)` needs to be `self.activaton(y).detach()`? Otherwise it will change the input value.

**Comment by guozhiyao** [comment]
And increase the amount of calculation。

**Comment by ZjjConan** [comment]
1. there is no need to do detach op. We also need to accumulate the gradient from activation(y) to update dydx for the previous bg. This is the same as most of the other attention modules, like SE, CBAM, and ECA.
2. In fact, detach may increase the time for training (I didn't test it), but not for testing. 

Hope it will help you

**Comment by guozhiyao** [comment]
It seems that this method will take up a lot of GPU memory during the training process. My original training process takes up 22G. After adding this method, it takes up more than 32G.

**Comment by guozhiyao** [comment]
Can I use the 'torch.utils.checkpoint.checkpoint' to reduce the GPU memory? During the `forward`, use the `checkpoint` to wrap the `attention_module`, like this
`
x = self.conv2(x)
x = checkpoint(self.attention_module, x)
`
and return the `x`.

**Comment by ZjjConan** [comment]
Hi @guozhiyao,
1. Actually, I usually use 4 cards for training and do not monitor the memory usage in my task. I think it will increase the memory usage for the training, but do not occupy too much memory for the test. 
2.  I didn't use the checkpoint method before. Sorry, can't help you.

**Comment by guozhiyao** [comment]
Thanks for your reply, I tried `checkpoint`, which can save a lot of GPU memory during training, and look forward to performance improvement.

**Comment by ZjjConan** [comment]
@guozhiyao Hope it can help you in your task. Thanks.
